torch/autograd/_functions/tensor.py

```diff
@@ -14,13 +14,19 @@ def forward(ctx, i, index):
         ctx.input_size = i.size()
         ctx.index = index
         result = i.index(ctx.index)
-        ctx.mark_shared_storage((i, result))
+        ctx.advanced_indexing = i._check_advanced_indexing(index)
+        if not ctx.advanced_indexing:
+            ctx.mark_shared_storage((i, result))
         return result
 
     @staticmethod
     def backward(ctx, grad_output):
-        grad_input = Variable(grad_output.data.new(ctx.input_size).zero_())
-        grad_input[ctx.index] = grad_output
+        grad_input = grad_output.data.new(ctx.input_size).zero_()
+        grad_input = Variable(grad_input)
+        if ctx.advanced_indexing:
+            grad_input._advanced_index_add(ctx.index, grad_output)
+        else:
+            grad_input[ctx.index] = grad_output
```

Comment:
Advanced indexing is only triggered if we have `ndim` sequence objects, where `ndim` is the number of dimensions of the Tensor being indexed. In subsequent PRs, with more advanced indexing functionality, this will be relaxed.